### PR TITLE
Fix problem with polytope sampler

### PR DESCRIPTION
**Pull request by jduerholt**
## Motivation

As discussed in issue https://github.com/pytorch/botorch/issues/1225, the polytope sampler runs into problems if the variables live in different dimensions. This PR fixes the issue using the approach mentioned in the issue above.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Unit tests

### CODE REVIEW

File: botorch/utils/sampling.py [hunk 1]
```diff
@@ -838,6 +861,7 @@ def get_polytope_samples(
     seed: Optional[int] = None,
     thinning: int = 32,
     n_burnin: int = 10_000,
+    normalize_bounds: bool = True,
```
Comment (Balandat) [inline botorch/utils/sampling.py:864]: Let's just do this by default and get rid of the option to simplify he interface? I don't really see a benefit for not doing this, worst case there is a bit of overhead but that's mostly negligible compared to the cost of drawing the samples.

File: test/utils/test_sampling.py [hunk 2]
```diff
@@ -289,6 +290,28 @@ def test_sparse_to_dense_constraints(self):
...
+            bounds = torch.tensor(
+                [[0.1, 0.3, 0.1, 30.0], [0.6, 0.7, 0.7, 700.0]], **tkwargs
+            )
+            new_constraints = normalize_linear_constraints(bounds, constraints)
+            expected_coefficients = torch.tensor([0.4000, 0.6000, 0.5000], **tkwargs)
+            self.assertTrue(
+                torch.allclose(new_constraints[0][1], expected_coefficients)
+            )
+            expected_rhs = 0.5
+            self.assertTrue(np.allclose(new_constraints[0][-1], expected_rhs))
```
Comment (Balandat) [inline test/utils/test_sampling.py:313]:
```suggestion
            self.assertAlmostEqual(new_constraints[0][-1], expected_rhs)
```

File: botorch/utils/sampling.py [hunk 3]
```diff
@@ -878,17 +907,26 @@ def get_polytope_samples(
...
     else:
         dense_equality_constraints = None
     polytope_sampler = HitAndRunPolytopeSampler(
         inequality_constraints=dense_inequality_constraints,
-        bounds=bounds,
+        bounds=torch.tensor(
+            [list(torch.zeros(bounds.shape[-1])), list(torch.ones(bounds.shape[-1]))]
+        ).to(bounds)
+        if normalize_bounds
+        else bounds,
```
Comment (Balandat) [inline botorch/utils/sampling.py:922]: would do sth like this for better readability:
```
normalized_bounds  = torch.zeros_like(bounds)
normalized_bounds[1, :] = 1.0
polytope_sampler = HitAndRunPolytopeSampler(...
```

**Review by Balandat — COMMENTED** [review]
Thanks for the PR, overall this looks great. My main comment is that we should just do this by default, no need to give the user the option to shoot themselves in the foot (unless there are any downsides to this that I'm not seeing right now).

File: botorch/utils/sampling.py [hunk 4]
```diff
@@ -830,6 +830,29 @@ def draw(self, n: int = 1, seed: Optional[int] = None) -> Tensor:
...
+            `\sum_i (X[indices[i]] * coefficients[i]) >= rhs` or
+            `\sum_i (X[indices[i]] * coefficients[i]) = rhs`.
+    """
+
+    new_constraints = []
+    for c in constraints:
+        lower = bounds[0, c[0].to(torch.int64)]
+        upper = bounds[1, c[0].to(torch.int64)]
+        s = upper - lower
+        new_constraints.append((c[0], s * c[1], float(c[2] - torch.sum(c[1] * lower))))
```
Comment (Balandat) [inline botorch/utils/sampling.py:852]:
```suggestion
        new_constraints.append((c[0], s * c[1], float(c[2] - torch.dot(c[1], lower))))
```

File: botorch/utils/sampling.py [hunk 5]
```diff
@@ -867,7 +894,9 @@ def get_polytope_samples(
     if inequality_constraints:
         A, b = sparse_to_dense_constraints(
             d=bounds.shape[-1],
-            constraints=inequality_constraints,
+            constraints=normalize_linear_constraints(bounds, inequality_constraints)
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/utils/sampling.py]: Done ;)

File: botorch/utils/sampling.py [hunk 6]
```diff
@@ -830,6 +830,29 @@ def draw(self, n: int = 1, seed: Optional[int] = None) -> Tensor:
...
+        bounds (Tensor): A `2 x d`-dim tensor containing the box bounds.
+        constraints (List[Tuple[Tensor, Tensor, float]]): A list of
+            tuples (indices, coefficients, rhs), with each tuple encoding
+            an inequality constraint of the form
+            `\sum_i (X[indices[i]] * coefficients[i]) >= rhs` or
+            `\sum_i (X[indices[i]] * coefficients[i]) = rhs`.
+    """
+
+    new_constraints = []
+    for c in constraints:
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/utils/sampling.py]: good idea, I implemented it as suggest.

File: botorch/utils/sampling.py (hunk 1 above)
Comment (jduerholt) [reply botorch/utils/sampling.py]: I agree, I removed the keyword.

File: botorch/utils/sampling.py (hunk 3 above)
Comment (jduerholt) [reply botorch/utils/sampling.py]: Good idea.

File: botorch/utils/sampling.py (hunk 4 above)
Comment (jduerholt) [reply botorch/utils/sampling.py]: Done.

File: botorch/utils/sampling.py [hunk 7]
```diff
@@ -830,6 +830,32 @@ def draw(self, n: int = 1, seed: Optional[int] = None) -> Tensor:
...
+            tuples (indices, coefficients, rhs), with each tuple encoding
+            an inequality constraint of the form
+            `\sum_i (X[indices[i]] * coefficients[i]) >= rhs` or
+            `\sum_i (X[indices[i]] * coefficients[i]) = rhs`.
+    """
+
+    new_constraints = []
+    for index, coefficient, rhs in constraints:
+        # the index is cast to int64 since in some tests the index is
+        # given as a indexing incompatible dtype
```
Comment (Balandat) [inline botorch/utils/sampling.py:850]: It looks like you fixed this in the tests also? Can we get rid of this comment and the casting below then? I'd like to avoid a situation where someone is confusing the order or the `indices` and `coefficient` tensors and suddenly ends up with something that doesn't error out but produces garbage results.

Comment (jduerholt) [reply botorch/utils/sampling.py]: There was another wrong type definition in one of the intializer tests. I also corrected this and removed the explicit cast.

File: test/optim/test_initializers.py [hunk 8]
```diff
@@ -275,14 +275,14 @@ def test_gen_batch_initial_conditions_constraints(self):
             bounds = torch.tensor([[0, 0], [1, 1]], device=self.device, dtype=dtype)
             inequality_constraints = [
                 (
-                    torch.tensor([1], device=self.device, dtype=dtype),
+                    torch.tensor([1], device=self.device, dtype=torch.int64),
```
Comment (esantorella) [inline test/optim/test_initializers.py:278]:
```suggestion
                    torch.tensor([1], device=self.device, dtype=dtype),
```
Let me know if I'm missing something, but tests still pass for me with the old version so I'd rather leave it how it is

File: test/optim/test_initializers.py [hunk 9]
```diff
@@ -275,14 +275,14 @@ def test_gen_batch_initial_conditions_constraints(self):
...
                     torch.tensor([-4], device=self.device, dtype=dtype),
                     torch.tensor(-3, device=self.device, dtype=dtype),
                 )
             ]
             equality_constraints = [
                 (
-                    torch.tensor([0], device=self.device, dtype=dtype),
+                    torch.tensor([0], device=self.device, dtype=torch.int64),
```
Comment (esantorella) [inline test/optim/test_initializers.py:285]:
```suggestion
                    torch.tensor([0], device=self.device, dtype=dtype),
```

File: test/optim/test_initializers.py (hunk 8 above)
Comment (esantorella) [inline test/optim/test_initializers.py:278]: Hmm, so non-int dtypes used to be acceptable in `gen_batch_initial_conditions` and `get_polytope_samples`, but now a non-int will cause a failure in `normalize_linear_constraints`. I'd prefer not to break backwards compatibility, but this should be fine for now since it isn't breaking any tests. I'll add this to the issue.

**Review by esantorella — APPROVED** [review]
Thanks so much for finding the bug and for fixing it! This looks great to me.

File: test/optim/test_initializers.py (hunk 8 above)
Comment (Balandat) [reply test/optim/test_initializers.py:278]: I think it's fine to break backward-compatibility here; it's kind of dicey to use floats to index tensor positions - more likely than not code that was doing that didn't actually mean to do that (that was the case with a type bug in our test suite as well).
